### PR TITLE
CORS-3256: GCP destroy updates for CAPG

### PR DESCRIPTION
**Pull request by barbacbd**
** Delete global scoped resources that are created during CAPG installs. 
** Update: Delete Global Resources for BackEnd services and forwarding rules 
** Update: Firewall Rule filter using regular expressions 
** New: Delete Target TCP Proxies

### CODE REVIEW

File: pkg/destroy/gcp/targetTCPProxies.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+						},
+						Amount: 1,
+					}},
+				})
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to list backend services")
```
Comment (bfournie) [inline pkg/destroy/gcp/targetTCPProxies.go:48]:
```suggestion
		return nil, errors.Wrapf(err, "failed to list target tcp services")
```

File: pkg/destroy/gcp/backendservice.go [hunk 2]
```diff
@@ -4,60 +4,95 @@ import (
...
-	return o.listBackendServicesWithFilter(ctx, "items(name),nextPageToken", o.clusterIDFilter(), nil)
+func (o *ClusterUninstaller) listBackendServices(ctx context.Context, scope resourceScope) ([]cloudResource, error) {
+	return o.listBackendServicesWithFilter(ctx, "items(name),nextPageToken", o.clusterIDFilter(), nil, scope)
+}
+
+func createBackendResources(filterFunc func(*compute.BackendService) bool, list *compute.BackendServiceList) []cloudResource {
+	result := []cloudResource{}
+
+	for _, item := range list.Items {
+		if filterFunc == nil || filterFunc != nil && filterFunc(item) {
```
Comment (r4f4) [inline pkg/destroy/gcp/backendservice.go:22]:
```suggestion
		if filterFunc == nil || filterFunc(item) {
```

File: pkg/destroy/gcp/bucket.go [hunk 3]
```diff
@@ -52,7 +52,7 @@ func (o *ClusterUninstaller) listBucketsWithFilter(ctx context.Context, fields s
 }
 
 func (o *ClusterUninstaller) deleteBucket(ctx context.Context, item cloudResource) error {
-	o.Logger.Debugf("Deleting storate bucket %s", item.name)
+	o.Logger.Debugf("Deleting storage bucket %s", item.name)
```
Comment (r4f4) [inline pkg/destroy/gcp/bucket.go:55]: Belongs to a separate commit.

File: pkg/destroy/gcp/bucketobject.go [hunk 4]
```diff
@@ -31,7 +31,7 @@ func (o *ClusterUninstaller) listBucketObjects(ctx context.Context, bucket cloud
 }
 
 func (o *ClusterUninstaller) deleteBucketObject(ctx context.Context, bucket cloudResource, item cloudResource) error {
-	o.Logger.Debugf("Deleting storate object %s/%s", bucket.name, item.name)
+	o.Logger.Debugf("Deleting storage object %s/%s", bucket.name, item.name)
```
Comment (r4f4) [inline pkg/destroy/gcp/bucketobject.go:34]: Separate commit.

File: pkg/destroy/gcp/forwardingrule.go [hunk 5]
```diff
@@ -4,68 +4,103 @@ import (
...
-	return o.listForwardingRulesWithFilter(ctx, "items(name,region,loadBalancingScheme),nextPageToken", o.clusterIDFilter(), nil)
+func (o *ClusterUninstaller) listForwardingRules(ctx context.Context, scope resourceScope) ([]cloudResource, error) {
+	return o.listForwardingRulesWithFilter(ctx, "items(name,region,loadBalancingScheme),nextPageToken", o.clusterIDFilter(), nil, scope)
+}
+
+func createForwardingRuleResources(filterFunc func(*compute.ForwardingRule) bool, list *compute.ForwardingRuleList) []cloudResource {
+	result := []cloudResource{}
+
+	for _, item := range list.Items {
+		if filterFunc == nil || filterFunc != nil && filterFunc(item) {
```
Comment (r4f4) [inline pkg/destroy/gcp/forwardingrule.go:22]:
```suggestion
		if filterFunc == nil || filterFunc(item) {
```

File: pkg/destroy/gcp/forwardingrule.go [hunk 6]
```diff
@@ -4,68 +4,103 @@ import (
...
-					quota:    quota,
-				})
-			}
-		}
+		result = append(result, createForwardingRuleResources(filterFunc, list)...)
 		return nil
 	})
 	if err != nil {
-		return nil, errors.Wrapf(err, "failed to list forwarding rules")
+		return nil, errors.Wrapf(err, "failed to list global forwarding rules")
```
Comment (r4f4) [inline pkg/destroy/gcp/forwardingrule.go:86]:
```suggestion
		return nil, errors.Wrapf(err, "failed to list regional forwarding rules")
```

File: pkg/destroy/gcp/gcp.go [hunk 7]
```diff
@@ -173,6 +183,7 @@ func (o *ClusterUninstaller) destroyCluster() (bool, error) {
 		{name: "Target Pools", execute: o.destroyTargetPools},
 		{name: "Instance groups", execute: o.destroyInstanceGroups},
 		{name: "Forwarding rules", execute: o.destroyForwardingRules},
+		{name: "Target TCP Proxies", execute: o.destroyTargetTCPProxies},
```
Comment (r4f4) [inline pkg/destroy/gcp/gcp.go:186]: This could go to a new commit.

File: pkg/destroy/gcp/targetTCPProxies.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
+package gcp
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
```
Comment (r4f4) [inline pkg/destroy/gcp/targetTCPProxies.go:6]: Let's not introduce this package in new code.

File: pkg/destroy/gcp/targetTCPProxies.go [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+	ctx, cancel := context.WithTimeout(ctx, defaultTimeout)
+	defer cancel()
+	result := []cloudResource{}
+	req := o.computeSvc.TargetTcpProxies.List(o.ProjectID).Fields(googleapi.Field(fields))
+	if len(filter) > 0 {
+		req = req.Filter(filter)
+	}
+	err := req.Pages(ctx, func(list *compute.TargetTcpProxyList) error {
+		for _, item := range list.Items {
+			if filterFunc == nil || (filterFunc != nil && filterFunc(item)) {
```
Comment (r4f4) [inline pkg/destroy/gcp/targetTCPProxies.go:29]:
```suggestion
			if filterFunc == nil || filterFunc(item) {
```

File: pkg/destroy/gcp/targetTCPProxies.go (hunk 1 above)
Comment (r4f4) [inline pkg/destroy/gcp/targetTCPProxies.go:48]:
```suggestion
		return nil, fmt.Errorf("failed to list backend services: %w", err)
```

File: pkg/destroy/gcp/targetTCPProxies.go [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+func (o *ClusterUninstaller) deleteTargetTCPProxy(ctx context.Context, item cloudResource) error {
+	o.Logger.Debugf("Deleting target TCP Proxies %s", item.name)
+	ctx, cancel := context.WithTimeout(ctx, defaultTimeout)
+	defer cancel()
+	op, err := o.computeSvc.TargetTcpProxies.Delete(o.ProjectID, item.name).RequestId(o.requestID(item.typeName, item.name)).Context(ctx).Do()
+	if err != nil && !isNoOp(err) {
+		o.resetRequestID(item.typeName, item.name)
+		return errors.Wrapf(err, "failed to delete target TCP proxy %s", item.name)
```
Comment (r4f4) [inline pkg/destroy/gcp/targetTCPProxies.go:60]:
```suggestion
		return fmt.Errorf("failed to delete target TCP proxy %s: %w", item.name, err)
```

File: pkg/destroy/gcp/targetTCPProxies.go [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+	}
+	if op != nil && op.Status == "DONE" && isErrorStatus(op.HttpErrorStatusCode) {
+		o.resetRequestID(item.typeName, item.name)
+		return errors.Errorf("failed to delete target TCP proxy %s with error: %s", item.name, operationErrorMessage(op))
```
Comment (r4f4) [inline pkg/destroy/gcp/targetTCPProxies.go:64]:
```suggestion
		return fmt.Errorf("failed to delete target TCP proxy %s with error: %s: %w", item.name, operationErrorMessage(op), err)
```

File: pkg/destroy/gcp/targetTCPProxies.go [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+	}
+	items := o.insertPendingItems("targettcpproxy", found)
+	for _, item := range items {
+		err := o.deleteTargetTCPProxy(ctx, item)
+		if err != nil {
+			o.errorTracker.suppressWarning(item.key, err, o.Logger)
+		}
+	}
+	if items = o.getPendingItems("targettcpproxy"); len(items) > 0 {
+		return errors.Errorf("%d items pending", len(items))
```
Comment (r4f4) [inline pkg/destroy/gcp/targetTCPProxies.go:89]:
```suggestion
		return fmt.Errorf("%d items pending", len(items))
```

File: pkg/destroy/gcp/targetTCPProxies.go
Comment (r4f4) [inline pkg/destroy/gcp/targetTCPProxies.go:1]: This file should be introduced in its own commit.

**Comment by bfournie** [comment]
I tested this with CAPG by doing an `create cluster` followed an `destroy cluster`. I verified that all created resources were cleaned up and the destroy command completed successfully.

File: pkg/destroy/gcp/forwardingrule.go [hunk 13]
```diff
@@ -2,77 +2,113 @@ package gcp
...
-				})
-			}
-		}
+		result = append(result, createForwardingRuleResources(filterFunc, list)...)
 		return nil
 	})
+
 	if err != nil {
-		return nil, errors.Wrapf(err, "failed to list forwarding rules")
+		return nil, fmt.Errorf("failed to list regional forwarding rules. %w", err)
```
Comment (r4f4) [inline pkg/destroy/gcp/forwardingrule.go:87]:
```suggestion
		return nil, fmt.Errorf("failed to list regional forwarding rules: %w", err)
```

File: pkg/destroy/gcp/forwardingrule.go [hunk 14]
```diff
@@ -85,19 +121,22 @@ func (o *ClusterUninstaller) deleteForwardingRule(ctx context.Context, item clou
...
+	for _, scope := range []resourceScope{gcpGlobalResource, gcpRegionalResource} {
+		found, err := o.listForwardingRules(ctx, scope)
 		if err != nil {
-			o.errorTracker.suppressWarning(item.key, err, o.Logger)
+			return fmt.Errorf("failed to list forwarding rules: %w", err)
+		}
+		items := o.insertPendingItems("forwardingrule", found)
+		for _, item := range items {
+			err := o.deleteForwardingRule(ctx, item, scope)
+			if err != nil {
```
Comment (r4f4) [inline pkg/destroy/gcp/forwardingrule.go:132]:
```suggestion
			if err := o.deleteForwardingRule(ctx, item, scope); err != nil {
```

File: pkg/destroy/gcp/forwardingrule.go [hunk 15]
```diff
@@ -85,19 +121,22 @@ func (o *ClusterUninstaller) deleteForwardingRule(ctx context.Context, item clou
...
+				o.Logger.Errorf("error here: %w", err)
```
Comment (r4f4) [inline pkg/destroy/gcp/forwardingrule.go:133]:
```suggestion
				o.Logger.Errorf("error deleting forwarding rule %s: %w", item.name, err)
```

File: pkg/destroy/gcp/backendservice.go [hunk 16]
```diff
@@ -2,69 +2,100 @@ package gcp
...
+			req = req.Filter(filter)
+		}
+		err := req.Pages(ctx, func(list *compute.BackendServiceList) error {
+			result = append(result, createBackendResources(filterFunc, list)...)
+			return nil
+		})
+		if err != nil {
+			return nil, fmt.Errorf("failed to list global backend services: %w", err)
+		}
+		return result, err
```
Comment (r4f4) [inline pkg/destroy/gcp/backendservice.go:64]:
```suggestion
		return result, nil
```
to be consistent with the other similar changes.

**Review by r4f4 — COMMENTED** [review]
/lgtm

**Comment by patrickdillon** [comment]
/hold

destroy is failing to delete some resources and taking too long

File: pkg/destroy/gcp/backendservice.go [hunk 17]
```diff
@@ -2,69 +2,100 @@ package gcp
...
 	"github.com/openshift/installer/pkg/types/gcp"
 )
 
-func (o *ClusterUninstaller) listBackendServices(ctx context.Context) ([]cloudResource, error) {
-	return o.listBackendServicesWithFilter(ctx, "items(name),nextPageToken", o.clusterIDFilter(), nil)
+func (o *ClusterUninstaller) listBackendServices(ctx context.Context, scope resourceScope) ([]cloudResource, error) {
+	return o.listBackendServicesWithFilter(ctx, "items(name),nextPageToken", o.clusterIDFilter(), nil, scope)
+}
+
+func createBackendResources(filterFunc func(*compute.BackendService) bool, list *compute.BackendServiceList) []cloudResource {
```
Comment (patrickdillon) [inline pkg/destroy/gcp/backendservice.go:18]: should this be something like `findBackendResources` instead of `createBackendResources`?

File: pkg/destroy/gcp/gcp.go [hunk 18]
```diff
@@ -40,6 +42,14 @@ const (
...
 	// used for resources created by the Cluster API GCP provider.
 	capgProviderOwnedLabelFmt = "capg-cluster-%s"
+
+	// gcpGlobalResource is an identifier to indicate that the resource(s)
+	// that are being deleted are globally scoped.
+	gcpGlobalResource resourceScope = "global"
+
+	// gcpRegionalResource is an identifier to indicate that the resource(s)
+	// that are being deleted are regionally scoped.
+	gcpRegionalResource resourceScope = "regional"
```
Comment (patrickdillon) [inline pkg/destroy/gcp/gcp.go:52]: Instead of switching on global vs regional, perhaps we just always try to delete both? 

We generally need to be careful about how we delete resources, but if something is tagged with the infraID for the cluster it should be safe to destroy.

Comment (barbacbd) [reply pkg/destroy/gcp/gcp.go:52]: @patrickdillon If there is a way to delete both global and regional resources with a single call that would work. As far as I am aware there is not, so the deletion requires finding then deleting global or finding then deleting regional resources. This is the reason that I have opted to create these identifiers.

File: pkg/destroy/gcp/backendservice.go (hunk 17 above)
Comment (barbacbd) [reply pkg/destroy/gcp/backendservice.go:18]: @patrickdillon I am changing the function name but this function is not finding backend resources this is creating cloudResources (the internal struct for deletion).
